app/screens/global_threads/threads_list/threads_list.tsx

```diff
@@ -0,0 +1,170 @@
+    teamId: string;
+    teammateNameDisplay: string;
+    testID: string;
+    threads: ThreadModel[];
+    theme: Theme;
+    unreadsCount: number;
+};
+
+const styles = StyleSheet.create({
+    messagesContainer: {
+        flexGrow: 1,
+    },
+    loadingStyle: {
+        alignItems: 'center',
+        flex: 1,
+        justifyContent: 'center',
+        paddingTop: 16,
+        paddingBottom: 8,
+    },
+});
+
+const ThreadsList = ({
+    setTab,
+    tab,
+    serverUrl,
+    teamId,
+    teammateNameDisplay,
+    testID,
+    theme,
+    threads,
+    unreadsCount,
+}: Props) => {
+    const intl = useIntl();
+    const hasCalled = useRef(false);
+    const [isLoading, setIsLoading] = useState(true);
+    const [endReached, setEndReached] = useState(false);
+
+    const noThreads = !threads?.length;
+    const lastThread = threads?.length ? threads[threads.length - 1] : null;
+
+    useEffect(() => {
+        // this is to be called only when there are no threads
+        if (tab === 'all' && noThreads && !hasCalled.current) {
+            setIsLoading(true);
+            fetchThreads(serverUrl, teamId).finally(() => {
+                hasCalled.current = true;
+                setIsLoading(false);
+            });
+        } else {
+            setIsLoading(false);
+        }
+    }, [serverUrl, tab, noThreads, teamId]);
+
+    const listEmptyComponent = useMemo(() => {
+        if (isLoading) {
+            return (
+                <Loading
+                    color={theme.buttonBg}
+                    containerStyle={styles.loadingStyle}
+                />
+            );
+        }
+        return (
+            <EmptyState
+                intl={intl}
+                isUnreads={tab === 'unreads'}
+                theme={theme}
+            />
+        );
+    }, [theme, tab, isLoading]);
+
+    const listFooterComponent = useMemo(() => {
+        if (tab === 'unreads') {
+            return null;
+        }
+
+        if (endReached) {
+            return (
+                <EndOfList/>
+            );
+        } else if (isLoading) {
+            return (
+                <Loading
+                    color={theme.buttonBg}
+                    containerStyle={styles.loadingStyle}
+                />
+            );
+        }
+
+        return null;
+    }, [isLoading, tab, theme, endReached]);
+
+    const handleEndReached = useCallback(() => {
+        if (!lastThread || tab === 'unreads' || endReached) {
+            return;
+        }
+
+        const options = {
+            before: lastThread.id,
+            perPage: General.CRT_CHUNK_SIZE,
+        };
+
+        setIsLoading(true);
+        fetchThreads(serverUrl, teamId, options).then((response) => {
+            if ('data' in response) {
+                setEndReached(response.data.threads.length < General.CRT_CHUNK_SIZE);
+            }
+        }).finally(() => {
+            setIsLoading(false);
+        });
+    }, [lastThread?.id, tab, endReached]);
```

Comment:
You also depend on teamId and serverUrl (even though those probably won't change).